bftengine/src/bcstatetransfer/Messages.hpp

```diff
@@ -40,55 +40,52 @@ class MsgType {
 };
 
 struct BCStateTranBaseMsg {
+  BCStateTranBaseMsg(uint16_t type) : type(type) {}
   uint16_t type;
+  // This struct and its derived structs are used to de/serialize message buffers sent over communication channels.
+  // Since virtual methods modify the memory layout of a struct, there cannot be any for both this base struct and its
+  // inheritors.
+};
+
+// VariableSizeMsg is useful for structs with a flexible array member
+template <typename T>
+class VariableSizeMsg {
+ public:
+  VariableSizeMsg(size_t dataSize) : buffer_(new concord::Byte[calcMsgSize(dataSize)]()) {}
+
+  T* operator->() const { return reinterpret_cast<T*>(buffer_.get()); }
+
+  char* getSerializedMsg() const { return reinterpret_cast<char*>(buffer_.get()); }
+
+  static size_t calcMsgSize(size_t dataSize) { return sizeof(T) - 1 + dataSize; }
+
+ private:
+  std::unique_ptr<concord::Byte[]> buffer_;
 };
 
 struct AskForCheckpointSummariesMsg : public BCStateTranBaseMsg {
-  AskForCheckpointSummariesMsg() {
-    memset(this, 0, sizeof(AskForCheckpointSummariesMsg));
-    type = MsgType::AskForCheckpointSummaries;
-  }
+  AskForCheckpointSummariesMsg() : BCStateTranBaseMsg{MsgType::AskForCheckpointSummaries} {}
 
-  uint64_t msgSeqNum;
-  uint64_t minRelevantCheckpointNum;
+  uint64_t msgSeqNum{};
```

Comment:
I personally prefer initializing in the (only) constructor, as initializing this way requires checking two places in the code (You will always check the constructor first).